design.md

```diff
@@ -0,0 +1,138 @@
+
+### Headless Execution
+
+- The kernel must not include any references to JavaFX so that it can run in a headless environment.
+
+### Data Logging
+
+- Every connected device resource must be automatically available as a data source. The resources should be loaded from the robot config file.
+  - Sensors return their sensor value (e.g., potentiometer -> voltage, encoder -> ticks, IMU -> values of axes).
+  - Actuators return their control signal (e.g., servo -> angle, solenoid -> digital state, motor -> voltage). These are the values the user's program commands, not the measured value (which would be a sensor).
+- A script may add its own data sources via *variables*.
+  - A variable is addressed by name (a String).
+  - State changes of variables are tracked such that the current state of a variable can be read from another thread in parallel.
+- Data types that can be logged:
+  - Numerical types (e.g., int, long, float, double, byte, char, short, boolean).
+  - Strings.
+  - Frame transforms.
+  - Images (e.g., frames from a camera feed).
+  - Arbitrary bytes.
+- Data that has been logged must be able to be saved and exported to other formats (e.g., CSV) if the data type supports it (e.g., when exporting to a CSV you can export numerical data but not images).
+- The amount of data to buffer must be configurable via the settings.
+
+### Hardware
+
+- A robot can be loaded from its config file and either a connection method (if connecting to the hardware) or simulator plugin (if connecting to a simulator).
+  - ! RFC: Schema of a robot config file.
+    - Kinematics
+    - Device
+    - Device resources
+    - Scripts for body control & cad, limb control & cad, joint control
+    - Device resources in the robot config file need a flag that says whether they are allowed to fail discovery or not. By default, this is false, meaning that all device resources need to complete discovery without errors or else the entire discovery process fails. If it is true, and that device resources fails discovery, then the discovery process does not fail. This could be useful for competition robots where the user would rather the robot work with partial functionality rather than not work at all.
+- All hardware access must go through a robot (i.e., no direct access to hardware).
+  - Scripts would not work in a simulator if they could directly access the hardware.
+- We should use most of the hardware layer from the previous version of the kernel.
+
+#### RPC
+
+- We should use the Bowler RPC protocol from the previous version of the kernel.
+  - Need to add support for asking the device its capabilities so that the kernel can figure out if it needs to upload a new program.
+
+#### Dependency Management
+
+- The kernel must be able to download the toolchain for the device specified in the robot config file. This is because of the [usability requirement](#usability).
```

Comment:
Can we clarify what is meant by "toolchain" here? (I don't see it defined anywhere else in the document, and I'm not entirely clear on what is meant myself).

Reply:
By toolchain I mean, for example, the ESP-IDF toolchain that is used to compile programs for the ESP.